ENH: implement sliding_window_view

fanjin-z opened the pull request:
Fixes  #7753

Test Cases for PR #10771 :

Test Case 1
```
arr = np.arange(12).reshape(3,4)
print(arr)
shape = np.array([2,2])
sliding_window_view(arr, shape)

input:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]

output:
array([[[[ 0,  1],
         [ 4,  5]],

        [[ 1,  2],
         [ 5,  6]],

        [[ 2,  3],
         [ 6,  7]]],


       [[[ 4,  5],
         [ 8,  9]],

        [[ 5,  6],
         [ 9, 10]],

        [[ 6,  7],
         [10, 11]]]])
```

Test Case 2:
```
arr = np.arange(16).reshape(4,4)
print(arr)
shape = [2,3]
sliding_window_view(arr, shape)

input:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

output:
array([[[[ 0,  1,  2],
         [ 4,  5,  6]],

        [[ 1,  2,  3],
         [ 5,  6,  7]]],


       [[[ 4,  5,  6],
         [ 8,  9, 10]],

        [[ 5,  6,  7],
         [ 9, 10, 11]]],


       [[[ 8,  9, 10],
         [12, 13, 14]],

        [[ 9, 10, 11],
         [13, 14, 15]]]])
```
Test Case 3
```
arr = np.arange(16).reshape(4,4)
print(arr)
sliding_window_view(arr)

input:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

output:
array([[[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11],
         [12, 13, 14, 15]]]])
```

Commit:
implement sliding_window_view #7753
Test cases are shown in the issue page.

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,35 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
 
+def sliding_window_view(x, shape=None):
+    """
+    Create rolling window views of the 2D array with the given shape.
```
No reason to limit to 2D

Fnjn replied:
I know. I plan to implement beyond 2d and strides > 2 soon.

eric-wieser replied:
It looks like what you have already works fine for > 2

Fnjn replied:
Actually not. 

This is a 3D test case:
```
arr = np.arange(27).reshape(3,3,3)
print(arr)
shape = [2,2,2]
sliding_window_view(arr, shape)

input:
[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]

 [[18 19 20]
  [21 22 23]
  [24 25 26]]]

output:
array([[[[[[ 0,  1],
           [ 3,  4]],

          [[ 9, 10],
           [12, 13]]],


         [[[ 1,  2],
           [ 4,  5]],

          [[10, 11],
           [13, 14]]]],



        [[[[ 3,  4],
           [ 6,  7]],

          [[12, 13],
           [15, 16]]],


         [[[ 4,  5],
           [ 7,  8]],

          [[13, 14],
           [16, 17]]]]],




       [[[[[ 9, 10],
           [12, 13]],

          [[18, 19],
           [21, 22]]],


         [[[10, 11],
           [13, 14]],

          [[19, 20],
           [22, 23]]]],



        [[[[12, 13],
           [15, 16]],

          [[21, 22],
           [24, 25]]],


         [[[13, 14],
           [16, 17]],

          [[22, 23],
           [25, 26]]]]]])
```

You can see there is some repeated elements. The expected output dims should be (2, 2, 2)

eric-wieser replied:
That output looks correct to me - `(2, 2, 2, 2, 2, 2)` is the right shape for a sliding window of shape `(2, 2, 2)` over an array of shape `(3, 3, 3)`

Fnjn replied:
It should work on N(>3) dim.

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,35 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    """
+    Create rolling window views of the 2D array with the given shape.
+
+    .. warning:: This function has to be used with extreme care, see notes from 'as_strided'.
```
You should design this function so that this is not true. Perhaps just set the readonly flag on the returned array?

Fnjn replied:
You mean the warning? This warning comes from the  'as_strided', which I this function depends on. I don't know if it's a good idea to remove this warning?

eric-wieser replied:
Yes, I mean the warning. This warning is aimed at _you_, the caller of `as_strided`. You should be the one taking extreme care, so that the caller of `sliding_window_view` does not have to

Fnjn replied:
Add shape and step_size check. Warning can be safety removed.

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,35 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    x : ndarray
+        2D array to create rolling window views.
+    shape : sequence of int, optional
+        The shape of the new array. Defaults to ``x.shape``.
```
This is not accurate - it's the shape of the window, right?

This default doesn't seem useful

Fnjn replied:
I use the same default setting as it is in 'as_strided'. Any change suggestion?

eric-wieser replied:
Remove the default entirely

Fnjn replied:
Shape default removed.

eric-wieser commented:
Test cases should be part of this PR, to prove that they pass

Fnjn commented:
Test cases now added in PR.

eric-wieser commented:
I mean, the test cases should be in the source code itself. There's probably a `test_stride_tricks` file somewhere.

Commit:
Add Example Cases

Fnjn commented:
Sorry about that. This is my first contribution. So I need to create a new py file for the test cases?

eric-wieser commented:
You should be able to find an existing file somewhere

eric-wieser commented:
This one: https://github.com/numpy/numpy/blob/master/numpy/lib/tests/test_stride_tricks.py

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,72 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+       [[[ 8,  9, 10],
+         [12, 13, 14]],
+        [[ 9, 10, 11],
+         [13, 14, 15]]]])
```
Looks like you lost a level of indent here

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,72 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    strides = x.strides
+
+    view_shape = np.concatenate((o, shape), axis=0)
+    view_strides = np.concatenate((strides, strides), axis=0)
```
Here you can just do `strides + strides`

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,72 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    ----------
+    x : ndarray
+        2D array to create rolling window views.
+    shape : sequence of int, optional
```
May as well force this to be a tuple

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,72 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    view_shape = np.concatenate((o, shape), axis=0)
+    view_strides = np.concatenate((strides, strides), axis=0)
+    return np.lib.stride_tricks.as_strided(x ,view_shape, view_strides)
```
nit: comma in the wrong place

Fnjn replied:
fixed

shoyer commented:
I think this function really ought to have the signature `sliding_window_view(array, size, axis)`, per the interface discussed in #7753.

Fnjn commented:
I think the axis parameter does not have a very good interpretation in this case, especially on N(>2) dim. And if someone want to overlook sliding on particular dim, they can set the shape in that dim as the shape of input array of that dim.

eric-wieser commented:
> they can set the window shape in that dim as the shape of input array of that dim

Interestingly setting it to `1` works too

Commit:
Add step_size and N-dim support

Fnjn commented:
Just  add step_size. Now it also should work on N(>2) dimensions. Sorry for late reply, finished my final yesterday.

Commit:
Add shape and step_size check. Remove warning.

Commit:
Remove shape default
Add step_size default's description.

Commit:
Give proper parameter name 'step'

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,92 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        Array to create rolling window views.
+    shape : sequence of int
+        The shape of the new array.
+    step: sequence of int, optional
```
I don't think this argument is useful - it's meaning is clearer when done via slicing after the fact

Fnjn replied:
that description is a mistake. It should be corrected as 
shape : sequence of int
        The shape of the sliding window.

Thanks for point out.

Commit:
fix a parameter description mistake

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,92 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    shape : sequence of int
+        The shape of the sliding window.
+    step: sequence of int, optional
+        The sliding step of view window for each dimension of input. Defaults to 1 on all dimensions.
```
Reiterating my comment earlier - I don't think the `step` argument belongs here: https://github.com/numpy/numpy/issues/7753#issuecomment-374836132

Fnjn replied:
Personally, I think sliding_window(x, shape, step) is more efficient than sliding_window(x, shape)[::step], since the latter one requires creating view first, then take steps. This could be problematic in large input array. Also, the 'step' parameter (also mentioned as 'stepsize', 'stride') is suggested by the first comment([Here](https://github.com/numpy/numpy/issues/7753#issue-160868188). I think it would be handy to keep this optional parameter.

eric-wieser replied:
> which could be problematic in large input array

The cost of a view is constant, and does not depend on the `.shape` of the array - so "large" is irrelevant.

Fnjn replied:
> Furthermore, the step argument is ambiguous - does it specify the step within a window, or the step between windows?

In deep learning world(where I came from),  step within a window is usually called 'dilation', step between windows is usually called 'stride'. However, in numpy 'strides' usually means step **by bytes** in each dimension. I agree that the naming causes confusion here, but I think 'step' (or whatever it called) is used quite frequently. Maybe we can come up with a better name?

Fnjn replied:
> Let's discuss the step argument in the issue, not in the implementation.

@eric-wieser I clarify the document by adding ' window shifts' to this parameter description. I think this could avoid ambiguity and this parameter is quite useful. Do you still think we should remove this parameter?

Commit:
implement test function for sliding_window_view()

Commit:
implement test function for sliding_window_view()

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,92 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+         [ 9, 10]],
+
+        [[ 6,  7],
+         [10, 11]]]])
```
You've lost 4 spaces from the beginning of every line here

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,92 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+         [ 8,  9]],
+
+        [[ 6,  7],
+         [10, 11]]]])
```
Missing 4 spaces at the beginning of every line here too

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -325,6 +326,60 @@ def test_as_strided():
+                        [[ 2,  3],[ 6,  7]]],
+                        [[[ 4,  5],[ 8,  9]],
+                        [[ 5,  6],[ 9, 10]],
+                        [[ 6,  7],[10, 11]]]])
```
Much more readable as
```python
    expected = np.array([[[[ 0,  1], [ 4,  5]],
                          [[ 1,  2], [ 5,  6]],
                          [[ 2,  3], [ 6,  7]]],
                         [[[ 4,  5], [ 8,  9]],
                          [[ 5,  6], [ 9, 10]],
                          [[ 6,  7], [10, 11]]]])
```

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -325,6 +326,60 @@ def test_as_strided():
+    expected = np.array([None])
+    assert_array_equal(arr_view, expected)
+
+    arr = np.arange(12).reshape(3,4)
```
A more intuitive array for understanding the result is

```
i, j = np.ogrid[:3,:4]
arr = 10*i + j
```

Since then one digit is columns, and the other is rows

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -325,6 +326,60 @@ def test_as_strided():
+    assert_array_equal(arr_view, expected)
+
+    arr = np.arange(12).reshape(3,4)
+    shape = [2,2]
```
Shapes should be tuples in examples

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -325,6 +326,60 @@ def test_as_strided():
+                        [[[ 4,  5,  6],[ 8,  9, 10]],
+                        [[ 5,  6,  7],[ 9, 10, 11]]],
+                        [[[ 8,  9, 10],[12, 13, 14]],
+                        [[ 9, 10, 11],[13, 14, 15]]]])
```
Same comment about spacing

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -325,6 +326,60 @@ def test_as_strided():
+
+    arr = np.arange(12).reshape(3,4)
+    shape = [2,2]
+    step = [1,2]
```
Both of these should be tuples. Consider swaping out the `arange(12)` as above too

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -325,6 +326,60 @@ def test_as_strided():
+    step = [1,1,2]
+    shape = [2,2,2]
+    arr_view = sliding_window_view(x, shape, step)
+    expected = np.array([[[[[[ 0,  1],[ 4,  5]],[[12, 13],[16, 17]]],
```
Stick a line break here so that the first line matches the rest of them

Fnjn replied:
fixed

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,92 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    assert len(x.shape) == len(shape) == len(step),\
+    "view shape or step dimension doesn't match with input array "
+    assert np.all(shape > 0) and np.all(step > 0),\
+    "view shape or step cannot contain non-positive number"
```
Instead of using `assert`, you should use `raise ValueError`

Fnjn replied:
fixed

Commit:
Fix according to @eric-wieser comments

Commit:
Change arange to ogrid in Examples

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,94 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    view_shape = np.concatenate((o, shape), axis=0)
+    view_strides = np.concatenate((view_strides, strides), axis=0)
+    view = np.lib.stride_tricks.as_strided(x, view_shape, view_strides)
+    return np.squeeze(view)
```
This is a bad idea, as it makes it impossible to ask for a length-1 window.

Fnjn replied:
fixed

Commit:
remove np.squeeze on return line

Fnjn commented:
@eric-wieser Any other changes need to be done?

fnjn committed:
Clarify document to avoid parameter confusion.

seberg commented:
Make the default not writable (or writeable) please?

Commit:
add `writable` and more explanation in docs

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
resolve a write conflit

Fnjn commented:
@seberg  The build fails on 
"numpy.lib.tests.test__datasource.TestDataSourceOpen.test_ValidHTTP" with "616: OSError". Any idea?

eric-wieser commented:
(to restart tests)

seberg commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,102 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        The steps of window shifts for each dimension on input array at a time.
+        If given, must have same length as number of input array dimensions.
+        Defaults to 1 on all dimensions.
+    writeable : bool, optional
```
We called it readonly in the broadcast_to internal function, so maybe that is the better choice even?

seberg replied:
There was some discussion back then on the choice, do not remember though.

seberg commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,102 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    -------
+    view : ndarray
+        Sliding window views of `x`. view.shape = (x.shape - shape) // step + 1
+
```
There should likely be notes, or notes pointing to `as_strided`? Is this supposed to be more discoverable then `as_strided`? What about the see also section, etc.? Some of the start might be nicer in notes too.

seberg commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,102 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    if np.any(o <= 0):
+        raise ValueError('window shape cannot larger than input array shape')
+
+    strides = x.strides
```
I have a bit of a bad code smell considering that `x` is never sanitized?

eric-wieser replied:
What sanitization are you after?

seberg commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,102 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    if step is None:
+        step = np.ones(len(x.shape), np.int)
+    else:
+        step = np.array(step, np.int)
```
Big no go, this has to be `np.intp`

seberg reviewed (commented):
Was the list ever polled about this, I do not wish to continue review really, but to me it sounds like a bit more thought might be necessary before putting this in. What about a `subok` flag like in `broadcast_to`, where is this function to be discoverable at? Only in this non-discoverable submodule? `as_strided` is hidden because it is inherently unsafe, like `broadcast_to` this function is (hopefully) safe....

Commit:
fixes according to @seberg review

Fnjn commented:
@seberg Thank you for your review.

The dangers of  `as_strided` are possible access to invalid memory and multi-write operations on overlapping memory. The former danger could be avoided by border check, which have been done in `sliding_window_view`. The resolution to the latter on is quite tricky, one solution might be instead of using `numpy.ndarray.view`, we use `numpy.copy` on creating return array. The drawback is that this method takes quite alot of memory, if modification of the result array is not needed.


Also, using broadcast_to doesn't help, because it also uses `numpy.ndarray.view` on creating result array. 

My suggestion is take the current method while `writeable` is False (default). Otherwise, we use numpy.copy instead, as mentioned above. What do you think?


p.s. `writeable` and `readonly` are both used in numpy, I think both names are easy to understand.

seberg commented:
I am not sure what you are pointing at, I mentioned broadcast_to 1. because it serves a very similar purpose, and 2. because we made it a top level function that is actually discoverable (unlike `as_strided` which is hidden away in a dark corner for people to find and think "oh what a nice hammer this makes" ;)). Ì believe we had a discussion about prefering `readonly=True` in `broadcast_to` even if in the end we decided to not (yet) make it part of the public API and force it upon users.

Fnjn commented:
What I meant is, both `broadcast_to` and `as_strided` return view, not copies of input array. And I think this is the cause of write hazard mentioned in `as_strided` warning notes.

Commit:
resolve write hazard

Commit:
remove outdated docs.

Fnjn commented:
Changes have made to resolve hazards mentioned in `as_strided`. Now, `sliding_window_view` is safe to use. To be honest,  given so many hours I have spent on it, I'm reluctant to rewrite the whole function for switching `as_strided` to `broadcast_to`, since final functionality is the same.

mattip commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,145 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    views is readonly. Bear in mind, return copies (writeable=True), could possibly
+    take memory multiple amount of origin array, due to overlapping windows.
+
+    For some cases, there may be more efficient approaches, such as FFT based algo discussed in #7753.
```
I don't like referring to github discussions from within NumPy documentation. Perhaps this is better?

For some cases there may be more efficient approaches to calculate transformations across multi-dimensional arrays, for instance `scipy.signal.fftconvolve`, where combining the iterating step with the calculation itself while storing partial results can result in significant speedups.

mattip commented:
This seems like a nice addition, could it make 1.15 ?

seberg commented:
Yes, but I think the name (mostly of the readonly/writeable keyword arg) and function placement still needs to be finalized probably.

eric-wieser reviewed (dismissed):
Just for the record, I'm -1 on the `step` parameter for the reasons I gave [here](https://github.com/numpy/numpy/issues/7753#issuecomment-374836132)

Commit:
change referring according to @mattip.
change 'writeable' to 'readonly' as @seberg suggest.
remove 'step' as @eric-wieser request

Commit:
fix test minor error

Fnjn commented:
change referring according to @mattip
change 'writeable' to 'readonly' as @seberg suggest.
remove 'step' as @eric-wieser  request

Thanks for review.

eric-wieser commented:
@seberg, can you restate your position on `writable` vs `readonly`? I agree that the default should be non-writable, but I feel like we should pick the name with greater prevalence in numpy already, and my hunch is that the spelling of that is `writable`

eric-wieser committed:
DOC: Grammar fixes

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,112 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    view = as_strided(x, view_shape, view_strides, subok=subok, writeable=not readonly)
+
+    if not readonly:
+        return view.copy()
```
I don't like this - if the user wants a copy, just make them use `sliding_window_view(...).copy()`. That automatically sets the `writeable` flag on the result.

seberg commented:
I thought we had a discussion on the name in https://github.com/numpy/numpy/pull/5371 but I can't find the part about writ(e)able vs. readonly there. Maybe it was an earlier discussion. The fact that it is readonly in the internal `broadcast_to` function was what makes me slightly lean to readonly.

eric-wieser commented:
@sebert: I'd argue that the `writeable` argument in `as_strided` takes priority, as that's closer to visible in the public interface. It also matches the `.flags.writeable` name. Indeed, I don't think that argument should do anything other than set `ret_val.flags.writeable` - not the `copy` it currently does.

seberg commented:
Yeah, I am a bit confused myself now, but then you are right, writeable is more discoverable. Maybe one reason that it snuck in, is that the default would be False and not True. But that seems fine to me.

Of course it should not do a copy?! It is basically saying: I know I am doing crazy things, if I didn't I would do a copy ;). If it returns a copy, we should not have the argument at all.

eric-wieser commented:
Just checking you agreed with [this comment](https://github.com/numpy/numpy/pull/10771/files#r187812948)

Fnjn commented:
I'm ok with not making copy.  But I think we could keep a reminder on documents that user should write with caution. It's easy to forget (like I often do) the return is view.

eric-wieser commented:
>  It's easy to forget (like I often do) the return is view.

That's why the default is for the result to _not_ be writeable:

```python
x = sliding_window_view(...)
x[...] = 0  # error: x is not writeable
```

Perhaps the docs should be something like
```
writeable : bool
    When true, allow writing to the returned view. The default is false, as this should be used with caution: the returned view contains the same memory location multiple times, so writing to one location will cause others to change.
```

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -111,6 +111,112 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    Returns
+    -------
+    view : ndarray
+        Sliding window views (or copies) of `x`. view.shape = x.shape - shape + 1
```
This is not true. Instead
```python
view.shape == tuple(x_s - s + 1 for x_s, s in zip(x.shape, shape)) + shape
```

eric-wieser replied:
Or far more simply, `view.shape[i] = x.shape[i] - shape[i] + 1

eric-wieser committed:
STY: Add missing line break required by PEP8

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -112,6 +112,113 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    view_strides = strides
+
+    view_shape = np.concatenate((o, shape), axis=0)
+    view_strides = np.concatenate((view_strides, strides), axis=0)
```
I think these would both be faster as:
```python
view_shape = tuple(o) + tuple(shape)
view_strides = x.strides + x.strides
```

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -112,6 +112,113 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        raise ValueError('window shape cannot larger than input array shape')
+
+    if type(readonly) != bool:
+        raise TypeError('readonly must be a boolean')
```
I don't think this is necessary - let `as_strided` worry about validating that argument

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -112,6 +112,113 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        raise TypeError('`shape` must be a sequence of integer')
+    else:
+        if shape.ndim > 1:
+            raise ValueError('`shape` must be one-dimensional sequence of integer')
```
I think you want `!= 1` here

eric-wieser commented on numpy/lib/stride_tricks.py:
```diff
@@ -112,6 +112,113 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        if len(x.shape) != len(shape):
+            raise ValueError("`shape` length doesn't match with input array dimensions")
+        if np.any(shape <= 0):
+            raise ValueError('`shape` cannot contain non-positive value')
```
I think 0 should be allowed, even though it's not that useful - it's a natural generalization

shoyer commented:
It's worth comparing the implementation to the corresponding function from skimage: https://github.com/scikit-image/scikit-image/blob/a51ffde418161bd08b6325ec3128c9e54a31c4a0/skimage/util/shape.py#L107

skimage probably also has tests that are worth a look and/or copying (this should be OK, since it shares the same 3-clause BSD license as numpy).

Is it true that the input array needs to be continuous? (I haven't really thought about this)

eric-wieser commented:
> Is it true that the input array needs to be continuous? 

No, any input array will work - even one that was already the result of sliding_window_view!

shoyer commented:
skimage also has [`view_as_blocks`](https://github.com/scikit-image/scikit-image/blob/a51ffde418161bd08b6325ec3128c9e54a31c4a0/skimage/util/shape.py#L10) (which might also be worth copying to NumPy!). I wonder if the requirement is copied over from there.

stefanv commented:
Here's the version we've been using in skimage: https://github.com/scikit-image/scikit-image/blob/master/skimage/util/shape.py

We specifically only apply this functionality on contiguous arrays, because the memory consumption can be enormous when copying.

stefanv commented:
Sorry, I see @shoyer just mentioned that version.  And @eric-wieser, trying to understand your comment here: how can you guarantee that a copy is not made if you allow non-contiguous data?

seberg commented:
Heh, I actually do not see why you have the copy if you have a non-contiguous array in skimage. The function works on all dimensions separately (and most of those are not contiguous anyway). It seems like it might have been an oversight 6 years ago and never questioned?

stefanv commented:
@seberg Ah, if I understand correctly what you're saying: as long as you're manipulating existing array strides, and not recalculating the strides from scratch, you should be fine, and that condition can be relaxed.

seberg commented:
Yeah, I guess the first version might not have done that (sticking to existing strides within the limits of existing dimensions).

eric-wieser commented:
I can't think of any reason that any type of striding function would care about contiguity, short of being poorly written to use `x.itemsize` instead of something from `x.strides`.

mattip commented:
This PR seems to have stalled after much work.

Fnjn commented:
@mattip Thanks for reminding me the existence of this PR, I have completely forgotten. :dizzy_face:
 I will get back to it asap.

fanjin-z committed:
Merge branch 'master' of github.com:numpy/numpy

fanjin-z committed:
+ Change readonly parameter to writeable.
+ Update writeable description.
+ Fix a few parameter checks.
+ Other minor improvements.

Fnjn commented:
Is there any other updates needed?

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
 
+def sliding_window_view(x, shape, subok=False, writeable=False):
+    """
+    Creates sliding window views of the N dimensional array with the given window
```
This should be a single summary line, followed by a blank line.

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    ----------
+    x : array_like
+        Array to create sliding window views from.
+
```
no blank lines needed between parameters

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    Returns
+    -------
+    view : ndarray
+        Sliding window views of the array. (view.shape[i] = x.shape[i] - shape[i] + 1)
```
need double backticks around the code snippet for correct formatting

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    view : ndarray
+        Sliding window views of the array. (view.shape[i] = x.shape[i] - shape[i] + 1)
+
+    See also
```
`See Also` (capital A)

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    Examples
+    --------
+    >>> i, j = np.ogrid[:3,:4]
```
Would be nice to have a more understandable example. Say a 1-D one to start with. Also, to create a 2-D array it's better to use a more regularly used function, e.g. `np.arange(12).reshape((3, 4))`

eric-wieser replied:
I think the current example is a nice one, but agreed we should start with a 1d example

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+
+    """
+    # first convert input to array, possibly keeping subclass
+    x = np.array(x, copy=False, subok=subok)
```
`np.asanyarray` is the normal way to convert to an array while preserving subclasses.

eric-wieser replied:
It's done this way for consistency with `as_strided`

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+    except:
+        raise TypeError('`shape` must be a sequence of integer')
+    else:
+        if shape.ndim != 1:
```
normally `int` is allowed for `shape` as well. It's awkward to have to write `(5,)` when `5` would work too. Compare e.g. with `np.zeros`

rgommers commented on numpy/lib/stride_tricks.py:
```diff
@@ -113,6 +113,101 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        if np.any(shape < 0):
+            raise ValueError('`shape` cannot contain negative value')
+
+    o = np.array(x.shape) - shape  + 1 # output shape
```
`o` --> `output` (or some other name, please no one-letter `o`

eric-wieser replied:
`output_shape` would be best

rgommers commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -336,6 +336,33 @@ def test_as_strided():
     assert_array_equal([r] * 3, a_view)
 
+
+def test_sliding_window_view():
```
seems like tests with 1/2/3-D arrays at least are needed here.

also a test with non-default strides.

and in the issue there's a comment that this implementation is more numerically stable than the pandas rolling window implementation. would be nice if that could be shown/guaranteed with a test that's numerically challenging

eric-wieser replied:
What does "numerically stable" mean for an procedure that doesn't perform any arithmetic?

shoyer replied:
It’s certainky more numerically stable to calculate a rolling window mean separately for each window. Pandas uses an algorithm that involves keeping track of a single “current” sum for the rolling window, so every element gets added and then subtracted. Of course, this can be way faster....

rgommers reviewed (changes requested):
does this work the same as `skimage.util.view_as_windows`? that has a check on `flags.contiguous` that's not present here. EDIT: never mind, that's an issue in `skimage`

eric-wieser commented:
I'm starting to think we have the wrong API here.

Here's an alternate spelling that we could use:
```python
def sliding_window_view_v2(x, shape, axis):
    # add all the type checking and perhaps convenience shorthands as this PR has here
    w_ndim = len(shape)
    assert w_ndim == len(axis)
    
    out_strides = x.strides + tuple(x.strides[ax] for ax in axis)
    
    # note: same axis can be windowed repeatedly
    x_shape_trimmed = list(x.shape)
    for ax, dim in zip(axis, shape):
        x_shape_trimmed[ax] -= dim + 1
    out_shape = x_shape_trimmed + shape
    return as_strided(x, strides=out_strides, shape=out_shape)
```

Going through some use cases:

1. Getting a sliding window along all axes
  * `old_sliding(a, shape)`
  * `new_sliding(a, shape, range(a.ndim))`
2. Getting a sliding window along just one axis
  * `old_sliding(a, (1, 3)).squeeze(axis=-2)`
  * `new_sliding(a, (3,), axis=(-2,))`
3. Getting a sliding window along just one axis, when `a` is a stack of arrays
  * `old_sliding(a, (1,)*(a.ndim - 1) + (3,)).squeeze(axis=-2)`
  * `new_sliding(a, (3,), axis=(-2,))`
4. Getting a sliding window along the same axis twice
  * `old_sliding(old_sliding(a, (1, 3)).squeeze(axis=-2), (1, 4, 1)).squeeze(axis=-2)`
  * `new_sliding(a, (3, 4), axis=(-2, -2))`
5. Getting a sliding window with the window axes in a different order to the input stack:
  * `old_sliding(a,  (1,)*(a.ndim - 1) (3,)).swapaxes(-1, -2)
  * `new_sliding(a, (1, 3), axis=(-1, -2))`

Dan-Patterson commented:
Is this coming in an upcoming release?

seberg commented:
@Dan-Patterson nothing very specific. I suppose this mostly needs a champion to see things through. The main thing being coming to a decision for the final API. Which maybe means looking at the different suggestions here and writing a mail to the mailing list making a proposition and providing some background.

jni commented:
A minor comment about `new_sliding`: could the axis keyword argument not have a default None that would make the first case equivalent to `old_sliding`?

eric-wieser commented:
@jni: Yes, that seems like a good idea to me.

eric-wieser commented:
Contrastring `new_sliding` above to the skimage functions, they can be implemented as:
```python
def view_as_windows(arr_in, window_shape, step=1):
    if not isinstance(window_shape, tuple):
        window_shape = (window_shape,) * arr_in.ndim
    if not isinstance(step, tuple):
        step = (step,) * arr_in.ndim
    all_windows = new_sliding(arr_in, window_shape)
    return all_window[np.s_[...,] + tuple(np.s_[::s,] for s in step)]
```
and
```python
def view_as_blocks(arr_in, block_shape):
    # missing all the skimage sanity checks
    all_windows = new_sliding(arr_in, block_shape)
    return all_window[np.s_[...,] + tuple(np.s_[::s,] for s in block_shape)]
```

stefanv commented:
@eric-wieser I don't quite follow the first example. You generate the overlapping blocks from the non-overlapping blocks view?

eric-wieser commented:
`new_sliding` generates overlapping blocks as implemented above, I generate the non-overlapping ones by slicing.

stefanv commented:
This was the line I was looking at:

```
return all_window[np.s_[...,] + tuple(np.s_[::s,] for s in step)]
```

But, now I see, this is to introduce slices other than 1.

zklaus commented:
This seems so close to the finishing line. Can we push it over?

@Fnjn:
- Thanks for getting this all started! It seems there is some support for moving to the `new_sliding` API proposed by @eric-wieser from @rgommers and @stefanv . Do you have an opinion on that?
- Could you mark the finished discussions above as resolved to simplify the reading of this issue?

zklaus commented:
@eric-wieser, @rgommers, the original author @Fnjn seems to have lost interest.

Is it ok to push forward regardless? What are the missing steps?

rgommers commented:
@zklaus yes it's okay to pick this up and continue in a new PR. Please just make sure to keep the old commits intact, so authorship gets preserved.

zklaus commented:
I pick this up in the replacement PR #17394.

Fnjn commented:
@zklaus Thank you for picking up this PR. I don't really have time to work on this. Wish you luck!

zklaus commented:
Thanks, @Fnjn!

seberg commented:
Superseded by gh-17394, please do have a look if it seems good if you like. Thanks for this PR!
